descartes_light/solvers/include/descartes_light/solvers/ompl/impl/descartes_space.hpp

```diff
@@ -0,0 +1,424 @@
+  long unsigned int rung_diff = max_rung - min_rung;
+  int idx_diff = static_cast<int>(idx1) - static_cast<int>(idx2);
+
+  // If they are on the same rung give a cost of 1 rung_to_rung_dist_, this should be seen as further away than points on adjacent
+  // rungs, but closer than points separated by 1 rung in between
+  if (rung_diff == 0)
+    dist += rung_to_rung_dist_;
+
+  // If the rung difference is greater than 1, find the rung difference, multiply by the distance per rung
+  // Then add the distance between the indices to help with growing the trees randomly
+  else if (rung_diff > 1)
+    dist += rung_to_rung_dist_ * static_cast<double>(rung_diff) + IDX_TO_IDX_DIST * abs(idx_diff);
+
+  // If the rungs are adjacent then proceed with evaluating the cost of going between them
+  if (rung_diff == 1)
+  {
+    // If the min rung is the zeroth rung or the last added rung, then the cost should be 0. These states were added to give
+    // a valid start and goal for the OMPL problem and should not be accounted for in the cost.
+    if (min_rung == 0 || max_rung == ladder_rungs_.size() - 1)
+      dist = 0;
+    else
+    {
+      // Perform edge evaluation. Use the min_rung - 1 edge evaluator because there was an extra rung added to the graph to
+      // provide a valid start state for OMPL which is not accounted for in the edge evaluator.
+      std::pair<bool, FloatType> edge_res
+          = edge_eval_[static_cast<std::size_t>(min_rung - 1)]->evaluate(*graph_[ladder_rungs_[rung1][idx1]].sample.state,
+                                                                         *graph_[ladder_rungs_[rung2][idx2]].sample.state);
+
+      // If this is a valid connection then the cost is added or else the cost of 1 rung distance is given
+      if (edge_res.first)
+        dist += static_cast<double>(edge_res.second);
+      else
+        dist += rung_to_rung_dist_;
+    }
+  }
+  // Ensure that this isn't going to the last rung artificially added, then add the cost of the higher rung in the connection
+  if (max_rung != ladder_rungs_.size() - 1)
+  {
+    if (rung1 == max_rung)
+      dist += graph_[ladder_rungs_[rung1][idx1]].sample.cost;
+    else
+      dist += graph_[ladder_rungs_[rung2][idx2]].sample.cost;
+  }
+
+  // If a cost of zero was found, add the distance epsilon because OMPL requires that the distance between any 2 distinct states is >0
+  if (dist == 0)
+    dist += distance_epsilon_;
+  return dist;
+}
+
+template <typename FloatType>
+bool descartes_light::DescartesStateSpace<FloatType>::equalStates(const ompl::base::State *state1, const ompl::base::State *state2) const
+{
+  long unsigned int rung1 = state1->as<StateType>()->vertex.first;
+  long unsigned int idx1 = state1->as<StateType>()->vertex.second;
+  long unsigned int rung2 = state2->as<StateType>()->vertex.first;
+  long unsigned int idx2 = state2->as<StateType>()->vertex.second;
+  if (rung1 == rung2 && idx1 == idx2)
+    return true;
+  return false;
```

Comment:
I would define an equality operator for `StateType` instead

Reply:
I tried to do this and the operator definition didn't pass through when I tried to call `state1->as<StateType>() == state2->as<StateType>()`